Add Gloo TCP_TLS transport

pbelevich opened the pull request:
This PR makes the following changes to root CMakeLists.txt:
1. If `USE_GLOO` is `ON` then try to build Gloo with OpenSSL headers(!) and set `USE_GLOO_WITH_OPENSSL` to `ON`
2. If `USE_GLOO_WITH_OPENSSL` is `ON` and the OS is Linux then try to find OpenSSL 1.1.x
3. If `OPENSSL_FOUND` then set `USE_TCP_OPENSSL_LOAD` to `ON` which means that Gloo will use OpenSSL headers to build TLS_TCP transport that is neither linked statically nor dynamically to OpenSSL but [tries to load it during runtime](https://github.com/facebookincubator/gloo/blob/2e0fe467aac7ed96a829548ec917edf6be8b1d84/gloo/transport/tcp/tls/openssl.cc#L4-L16)

@malfet @seemethere **I assume that during the release all CI release Linux environments will have OpenSSL 1.1.1(openssl-devel/libssl-dev) installed and it will automatically build PyTorch with TLS_TCP transport support.** 

**The remaining issue: mobile builds and libtorch builds have OpenSSL 1.0.2 which is not compatible with my code**

Depens on https://github.com/pytorch/builder/pull/708

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56442 Add Gloo TCP_TLS transport**

Differential Revision: [D27896285](https://our.internmc.facebook.com/intern/diff/D27896285)

pbelevich committed:
Add Gloo TCP_TLS transport

[ghstack-poisoned]

pbelevich committed:
Update on "Add Gloo TCP_TLS transport"




[ghstack-poisoned]

pbelevich committed:
Update on "Add Gloo TCP_TLS transport"




[ghstack-poisoned]

pbelevich committed:
Update on "Add Gloo TCP_TLS transport"




[ghstack-poisoned]

pbelevich committed:
Update on "Add Gloo TCP_TLS transport"




[ghstack-poisoned]

pbelevich committed:
Update on "Add Gloo TCP_TLS transport"




[ghstack-poisoned]

pbelevich committed:
Update on "Add Gloo TCP_TLS transport"




[ghstack-poisoned]

pbelevich commented on .jenkins/pytorch/test.sh:
```diff
@@ -438,6 +448,7 @@ elif [[ "${BUILD_ENVIRONMENT}" == *-test1 || "${JOB_BASE_NAME}" == *-test1 ]]; t
   test_python_shard1
   test_aten
 elif [[ "${BUILD_ENVIRONMENT}" == *-test2 || "${JOB_BASE_NAME}" == *-test2 ]]; then
+  test_python_gloo_with_tls
```
put it at first place for testing purpose, will move it to the end after resolve all issues

malfet replied:
Please make sure to limit it to just few configs.

malfet replied:
To elaborate on my previous comment, invoke this function if BUILD_ENVIRONEMNT equals to `pytorch_linux_xenial_py3_6_gcc7_test`

pbelevich replied:
addressed

pbelevich commented on CMakeLists.txt:
```diff
@@ -324,6 +329,15 @@ if(WIN32)
 endif()
 
+if(USE_GLOO_WITH_OPENSSL AND LINUX)
+  find_package(OpenSSL REQUIRED)
```
Made it OPENSSL_init_ssl for testing purposes to make sure that we have libssl-dev/openssl-develop installed on CI environments

pbelevich committed:
Update on "Add Gloo TCP_TLS transport"


Differential Revision: [D27896285](https://our.internmc.facebook.com/intern/diff/D27896285)

[ghstack-poisoned]

pbelevich committed:
Update on "Add Gloo TCP_TLS transport"


Differential Revision: [D27896285](https://our.internmc.facebook.com/intern/diff/D27896285)

[ghstack-poisoned]

rohan-varma reviewed (commented):
Are we planning to document `GLOO_DEVICE_TRANSPORT=TCP_TLS` (can be done in future PRs) can be set to enable tls transport?

pbelevich committed:
Update on "Add Gloo TCP_TLS transport"


Differential Revision: [D27896285](https://our.internmc.facebook.com/intern/diff/D27896285)

[ghstack-poisoned]

pbelevich committed:
Update on "Add Gloo TCP_TLS transport"


Differential Revision: [D27896285](https://our.internmc.facebook.com/intern/diff/D27896285)

[ghstack-poisoned]

pbelevich committed:
Update on "Add Gloo TCP_TLS transport"


Differential Revision: [D27896285](https://our.internmc.facebook.com/intern/diff/D27896285)

[ghstack-poisoned]

malfet commented on .jenkins/pytorch/create_test_cert.sh:
```diff
@@ -0,0 +1,18 @@
+
+TMP_CERT_DIR=$(mktemp -d)
+
+openssl genrsa -out ${TMP_CERT_DIR}/ca.key 2048
+openssl req -subj "/C=US/ST=New York/L=New York/O=Gloo Certificate Authority" -new -x509 -days 7300 -key ${TMP_CERT_DIR}/ca.key -sha256 -extensions v3_ca -out ${TMP_CERT_DIR}/ca.pem
+
+openssl genrsa -out ${TMP_CERT_DIR}/pkey.key 2048
+openssl req -subj "/C=US/ST=California/L=San Francisco/O=Gloo Testing Company" -sha256 -new -key ${TMP_CERT_DIR}/pkey.key -out ${TMP_CERT_DIR}/csr.csr
+
+openssl x509 -sha256 -req -in ${TMP_CERT_DIR}/csr.csr -CA ${TMP_CERT_DIR}/ca.pem -CAkey ${TMP_CERT_DIR}/ca.key -CAcreateserial -out ${TMP_CERT_DIR}/cert.pem -days 7300
+
+openssl verify -CAfile ${TMP_CERT_DIR}/ca.pem ${TMP_CERT_DIR}/cert.pem
+
+export GLOO_DEVICE_TRANSPORT=TCP_TLS
+export GLOO_DEVICE_TRANSPORT_TCP_TLS_PKEY=${TMP_CERT_DIR}/pkey.key
+export GLOO_DEVICE_TRANSPORT_TCP_TLS_CERT=${TMP_CERT_DIR}/cert.pem
+export GLOO_DEVICE_TRANSPORT_TCP_TLS_CA_FILE=${TMP_CERT_DIR}/ca.pem
```
Hmm, in order to make it easier for external users to run, can we add this logic to respective test in `test_c10d_gloo` using cryptography package (see https://cryptography.io/ )?
Something like:
with TemporaryDirectory() as temp_dir:
   from cryptography import x509
   subject_name = x509.Name([x509.NameAttributed(COUNTRY_NAME, "VA"),...])
   x509.CertificateBuilder().subject_name(subject_name).sign().....

pbelevich replied:
addressed

malfet commented on torch/lib/c10d/GlooDeviceFactory.cpp:
```diff
@@ -59,6 +63,35 @@ C10_REGISTER_CREATOR(GlooDeviceRegistry, LINUX, makeTCPDevice);
 #endif
 
+#if GLOO_HAVE_TRANSPORT_TCP_TLS
+static std::string env_to_string(const char* chars) {
```
I think function name is a bit misleading isn't it?
```suggestion
static std::string cstr_to_std_string(const char* c_str) {
```

pbelevich replied:
addressed

malfet commented on torch/lib/c10d/GlooDeviceFactory.cpp:
```diff
@@ -59,6 +63,35 @@ C10_REGISTER_CREATOR(GlooDeviceRegistry, LINUX, makeTCPDevice);
 
+#if GLOO_HAVE_TRANSPORT_TCP_TLS
+static std::string env_to_string(const char* chars) {
+  return chars == nullptr ? std::string("") : std::string(chars);
```
nit
```suggestion
  return std::string (chars != nullptr ? chars : "");
```

pbelevich replied:
addressed

malfet commented:
LGTM, but please do not land until CI issues are resolved.

pbelevich committed:
Update on "Add Gloo TCP_TLS transport"

This PR makes the following changes to root CMakeLists.txt:
1. If `USE_GLOO` is `ON` then try to build Gloo with OpenSSL headers(!) and set `USE_GLOO_WITH_OPENSSL` to `ON`
2. If `USE_GLOO_WITH_OPENSSL` is `ON` and the OS is Linux then try to find OpenSSL 1.1.x
3. If `OPENSSL_FOUND` then set `USE_TCP_OPENSSL_LOAD` to `ON` which means that Gloo will use OpenSSL headers to build TLS_TCP transport that is neither linked statically nor dynamically to OpenSSL but [tries to load it during runtime](https://github.com/facebookincubator/gloo/blob/2e0fe467aac7ed96a829548ec917edf6be8b1d84/gloo/transport/tcp/tls/openssl.cc#L4-L16)

@malfet @seemethere **I assume that during the release all CI release Linux environments will have OpenSSL 1.1.1(openssl-devel/libssl-dev) installed and it will automatically build PyTorch with TLS_TCP transport support.** 

**The remaining issue: mobile builds and libtorch builds have OpenSSL 1.0.2 which is not compatible with my code**

Depens on https://github.com/pytorch/builder/pull/707


Differential Revision: [D27896285](https://our.internmc.facebook.com/intern/diff/D27896285)

[ghstack-poisoned]

seemethere commented:
I've went ahead and added `/opt/openssl` to the CI docker images so that this PR can pass the tests. Once that PR is merge you can rebase and re-run the test for these

malfet commented on CMakeLists.txt:
```diff
@@ -324,6 +329,15 @@ if(WIN32)
 endif()
 
+if(USE_GLOO_WITH_OPENSSL AND LINUX)
+  find_package(OpenSSL 1.1 REQUIRED EXACT)
```
Please remove required before merging the PR. Also, can you add a comment about why EXACT is needed here?

malfet commented on CMakeLists.txt:
```diff
@@ -276,6 +276,11 @@ cmake_dependent_option(
     "USE_DISTRIBUTED" OFF)
+# If USE_GLOO then try to build Gloo with OpenSSL(currently available only on Linux).
+# set USE_GLOO_WITH_OPENSSL=OFF if you don't have OpenSSL or don't want to use it in Gloo
+cmake_dependent_option(
+    USE_GLOO_WITH_OPENSSL "Use Gloo with OpenSSL. Only available if USE_GLOO is on." ON
+    "USE_GLOO" OFF)
```
Can you please remove `AND LINUX` check from line 332 and move it here
```suggestion
cmake_dependent_option(
    USE_GLOO_WITH_OPENSSL "Use Gloo with OpenSSL. Only available if USE_GLOO is on." ON
    "USE_GLOO AND LINUX" OFF)
```

malfet commented on .jenkins/pytorch/create_test_cert.sh:
```diff
@@ -0,0 +1,18 @@
+
+openssl verify -CAfile ${TMP_CERT_DIR}/ca.pem ${TMP_CERT_DIR}/cert.pem
+
+export GLOO_DEVICE_TRANSPORT=TCP_TLS
```
Regardless of this test, can we please have a test that setting `GLOO_DEVICE_TRANSPORT` to `FOOBAR` raises an exception (this way we can rely that this test will fail if PyTorch was compiled without TCP_TLS support)

malfet commented on .jenkins/pytorch/test.sh:
```diff
@@ -149,6 +149,16 @@ test_python() {
 
+test_python_gloo_with_tls() {
+  source ./create_test_cert.sh
+  time python test/run_test.py --include test_c10d_gloo --verbose --determine-from="$DETERMINE_FROM"
```
Can we just run one or two subtests from the testsuite rather than a full one?

pbelevich committed:
Update on "Add Gloo TCP_TLS transport"

This PR makes the following changes to root CMakeLists.txt:
1. If `USE_GLOO` is `ON` then try to build Gloo with OpenSSL headers(!) and set `USE_GLOO_WITH_OPENSSL` to `ON`
2. If `USE_GLOO_WITH_OPENSSL` is `ON` and the OS is Linux then try to find OpenSSL 1.1.x
3. If `OPENSSL_FOUND` then set `USE_TCP_OPENSSL_LOAD` to `ON` which means that Gloo will use OpenSSL headers to build TLS_TCP transport that is neither linked statically nor dynamically to OpenSSL but [tries to load it during runtime](https://github.com/facebookincubator/gloo/blob/2e0fe467aac7ed96a829548ec917edf6be8b1d84/gloo/transport/tcp/tls/openssl.cc#L4-L16)

@malfet @seemethere **I assume that during the release all CI release Linux environments will have OpenSSL 1.1.1(openssl-devel/libssl-dev) installed and it will automatically build PyTorch with TLS_TCP transport support.** 

**The remaining issue: mobile builds and libtorch builds have OpenSSL 1.0.2 which is not compatible with my code**

Depens on https://github.com/pytorch/builder/pull/708


Differential Revision: [D27896285](https://our.internmc.facebook.com/intern/diff/D27896285)

[ghstack-poisoned]

pbelevich committed:
Update on "Add Gloo TCP_TLS transport"

This PR makes the following changes to root CMakeLists.txt:
1. If `USE_GLOO` is `ON` then try to build Gloo with OpenSSL headers(!) and set `USE_GLOO_WITH_OPENSSL` to `ON`
2. If `USE_GLOO_WITH_OPENSSL` is `ON` and the OS is Linux then try to find OpenSSL 1.1.x
3. If `OPENSSL_FOUND` then set `USE_TCP_OPENSSL_LOAD` to `ON` which means that Gloo will use OpenSSL headers to build TLS_TCP transport that is neither linked statically nor dynamically to OpenSSL but [tries to load it during runtime](https://github.com/facebookincubator/gloo/blob/2e0fe467aac7ed96a829548ec917edf6be8b1d84/gloo/transport/tcp/tls/openssl.cc#L4-L16)

@malfet @seemethere **I assume that during the release all CI release Linux environments will have OpenSSL 1.1.1(openssl-devel/libssl-dev) installed and it will automatically build PyTorch with TLS_TCP transport support.** 

**The remaining issue: mobile builds and libtorch builds have OpenSSL 1.0.2 which is not compatible with my code**

Depens on https://github.com/pytorch/builder/pull/708


Differential Revision: [D27896285](https://our.internmc.facebook.com/intern/diff/D27896285)

[ghstack-poisoned]

pbelevich committed:
Update on "Add Gloo TCP_TLS transport"

This PR makes the following changes to root CMakeLists.txt:
1. If `USE_GLOO` is `ON` then try to build Gloo with OpenSSL headers(!) and set `USE_GLOO_WITH_OPENSSL` to `ON`
2. If `USE_GLOO_WITH_OPENSSL` is `ON` and the OS is Linux then try to find OpenSSL 1.1.x
3. If `OPENSSL_FOUND` then set `USE_TCP_OPENSSL_LOAD` to `ON` which means that Gloo will use OpenSSL headers to build TLS_TCP transport that is neither linked statically nor dynamically to OpenSSL but [tries to load it during runtime](https://github.com/facebookincubator/gloo/blob/2e0fe467aac7ed96a829548ec917edf6be8b1d84/gloo/transport/tcp/tls/openssl.cc#L4-L16)

@malfet @seemethere **I assume that during the release all CI release Linux environments will have OpenSSL 1.1.1(openssl-devel/libssl-dev) installed and it will automatically build PyTorch with TLS_TCP transport support.** 

**The remaining issue: mobile builds and libtorch builds have OpenSSL 1.0.2 which is not compatible with my code**

Depens on https://github.com/pytorch/builder/pull/708


Differential Revision: [D27896285](https://our.internmc.facebook.com/intern/diff/D27896285)

[ghstack-poisoned]

pbelevich committed:
Update on "Add Gloo TCP_TLS transport"

This PR makes the following changes to root CMakeLists.txt:
1. If `USE_GLOO` is `ON` then try to build Gloo with OpenSSL headers(!) and set `USE_GLOO_WITH_OPENSSL` to `ON`
2. If `USE_GLOO_WITH_OPENSSL` is `ON` and the OS is Linux then try to find OpenSSL 1.1.x
3. If `OPENSSL_FOUND` then set `USE_TCP_OPENSSL_LOAD` to `ON` which means that Gloo will use OpenSSL headers to build TLS_TCP transport that is neither linked statically nor dynamically to OpenSSL but [tries to load it during runtime](https://github.com/facebookincubator/gloo/blob/2e0fe467aac7ed96a829548ec917edf6be8b1d84/gloo/transport/tcp/tls/openssl.cc#L4-L16)

@malfet @seemethere **I assume that during the release all CI release Linux environments will have OpenSSL 1.1.1(openssl-devel/libssl-dev) installed and it will automatically build PyTorch with TLS_TCP transport support.** 

**The remaining issue: mobile builds and libtorch builds have OpenSSL 1.0.2 which is not compatible with my code**

Depens on https://github.com/pytorch/builder/pull/708


Differential Revision: [D27896285](https://our.internmc.facebook.com/intern/diff/D27896285)

[ghstack-poisoned]

pbelevich committed:
Update on "Add Gloo TCP_TLS transport"

This PR makes the following changes to root CMakeLists.txt:
1. If `USE_GLOO` is `ON` then try to build Gloo with OpenSSL headers(!) and set `USE_GLOO_WITH_OPENSSL` to `ON`
2. If `USE_GLOO_WITH_OPENSSL` is `ON` and the OS is Linux then try to find OpenSSL 1.1.x
3. If `OPENSSL_FOUND` then set `USE_TCP_OPENSSL_LOAD` to `ON` which means that Gloo will use OpenSSL headers to build TLS_TCP transport that is neither linked statically nor dynamically to OpenSSL but [tries to load it during runtime](https://github.com/facebookincubator/gloo/blob/2e0fe467aac7ed96a829548ec917edf6be8b1d84/gloo/transport/tcp/tls/openssl.cc#L4-L16)

@malfet @seemethere **I assume that during the release all CI release Linux environments will have OpenSSL 1.1.1(openssl-devel/libssl-dev) installed and it will automatically build PyTorch with TLS_TCP transport support.** 

**The remaining issue: mobile builds and libtorch builds have OpenSSL 1.0.2 which is not compatible with my code**

Depens on https://github.com/pytorch/builder/pull/708


Differential Revision: [D27896285](https://our.internmc.facebook.com/intern/diff/D27896285)

[ghstack-poisoned]

pbelevich committed:
Update on "Add Gloo TCP_TLS transport"

This PR makes the following changes to root CMakeLists.txt:
1. If `USE_GLOO` is `ON` then try to build Gloo with OpenSSL headers(!) and set `USE_GLOO_WITH_OPENSSL` to `ON`
2. If `USE_GLOO_WITH_OPENSSL` is `ON` and the OS is Linux then try to find OpenSSL 1.1.x
3. If `OPENSSL_FOUND` then set `USE_TCP_OPENSSL_LOAD` to `ON` which means that Gloo will use OpenSSL headers to build TLS_TCP transport that is neither linked statically nor dynamically to OpenSSL but [tries to load it during runtime](https://github.com/facebookincubator/gloo/blob/2e0fe467aac7ed96a829548ec917edf6be8b1d84/gloo/transport/tcp/tls/openssl.cc#L4-L16)

@malfet @seemethere **I assume that during the release all CI release Linux environments will have OpenSSL 1.1.1(openssl-devel/libssl-dev) installed and it will automatically build PyTorch with TLS_TCP transport support.** 

**The remaining issue: mobile builds and libtorch builds have OpenSSL 1.0.2 which is not compatible with my code**

Depens on https://github.com/pytorch/builder/pull/708


Differential Revision: [D27896285](https://our.internmc.facebook.com/intern/diff/D27896285)

[ghstack-poisoned]

pbelevich committed:
Update on "Add Gloo TCP_TLS transport"

This PR makes the following changes to root CMakeLists.txt:
1. If `USE_GLOO` is `ON` then try to build Gloo with OpenSSL headers(!) and set `USE_GLOO_WITH_OPENSSL` to `ON`
2. If `USE_GLOO_WITH_OPENSSL` is `ON` and the OS is Linux then try to find OpenSSL 1.1.x
3. If `OPENSSL_FOUND` then set `USE_TCP_OPENSSL_LOAD` to `ON` which means that Gloo will use OpenSSL headers to build TLS_TCP transport that is neither linked statically nor dynamically to OpenSSL but [tries to load it during runtime](https://github.com/facebookincubator/gloo/blob/2e0fe467aac7ed96a829548ec917edf6be8b1d84/gloo/transport/tcp/tls/openssl.cc#L4-L16)

@malfet @seemethere **I assume that during the release all CI release Linux environments will have OpenSSL 1.1.1(openssl-devel/libssl-dev) installed and it will automatically build PyTorch with TLS_TCP transport support.** 

**The remaining issue: mobile builds and libtorch builds have OpenSSL 1.0.2 which is not compatible with my code**

Depens on https://github.com/pytorch/builder/pull/708


Differential Revision: [D27896285](https://our.internmc.facebook.com/intern/diff/D27896285)

[ghstack-poisoned]

pbelevich committed:
Update on "Add Gloo TCP_TLS transport"

This PR makes the following changes to root CMakeLists.txt:
1. If `USE_GLOO` is `ON` then try to build Gloo with OpenSSL headers(!) and set `USE_GLOO_WITH_OPENSSL` to `ON`
2. If `USE_GLOO_WITH_OPENSSL` is `ON` and the OS is Linux then try to find OpenSSL 1.1.x
3. If `OPENSSL_FOUND` then set `USE_TCP_OPENSSL_LOAD` to `ON` which means that Gloo will use OpenSSL headers to build TLS_TCP transport that is neither linked statically nor dynamically to OpenSSL but [tries to load it during runtime](https://github.com/facebookincubator/gloo/blob/2e0fe467aac7ed96a829548ec917edf6be8b1d84/gloo/transport/tcp/tls/openssl.cc#L4-L16)

@malfet @seemethere **I assume that during the release all CI release Linux environments will have OpenSSL 1.1.1(openssl-devel/libssl-dev) installed and it will automatically build PyTorch with TLS_TCP transport support.** 

**The remaining issue: mobile builds and libtorch builds have OpenSSL 1.0.2 which is not compatible with my code**

Depens on https://github.com/pytorch/builder/pull/708


Differential Revision: [D27896285](https://our.internmc.facebook.com/intern/diff/D27896285)

[ghstack-poisoned]

pbelevich committed:
Update on "Add Gloo TCP_TLS transport"

This PR makes the following changes to root CMakeLists.txt:
1. If `USE_GLOO` is `ON` then try to build Gloo with OpenSSL headers(!) and set `USE_GLOO_WITH_OPENSSL` to `ON`
2. If `USE_GLOO_WITH_OPENSSL` is `ON` and the OS is Linux then try to find OpenSSL 1.1.x
3. If `OPENSSL_FOUND` then set `USE_TCP_OPENSSL_LOAD` to `ON` which means that Gloo will use OpenSSL headers to build TLS_TCP transport that is neither linked statically nor dynamically to OpenSSL but [tries to load it during runtime](https://github.com/facebookincubator/gloo/blob/2e0fe467aac7ed96a829548ec917edf6be8b1d84/gloo/transport/tcp/tls/openssl.cc#L4-L16)

@malfet @seemethere **I assume that during the release all CI release Linux environments will have OpenSSL 1.1.1(openssl-devel/libssl-dev) installed and it will automatically build PyTorch with TLS_TCP transport support.** 

**The remaining issue: mobile builds and libtorch builds have OpenSSL 1.0.2 which is not compatible with my code**

Depens on https://github.com/pytorch/builder/pull/708


Differential Revision: [D27896285](https://our.internmc.facebook.com/intern/diff/D27896285)

[ghstack-poisoned]

pbelevich committed:
Update on "Add Gloo TCP_TLS transport"

This PR makes the following changes to root CMakeLists.txt:
1. If `USE_GLOO` is `ON` then try to build Gloo with OpenSSL headers(!) and set `USE_GLOO_WITH_OPENSSL` to `ON`
2. If `USE_GLOO_WITH_OPENSSL` is `ON` and the OS is Linux then try to find OpenSSL 1.1.x
3. If `OPENSSL_FOUND` then set `USE_TCP_OPENSSL_LOAD` to `ON` which means that Gloo will use OpenSSL headers to build TLS_TCP transport that is neither linked statically nor dynamically to OpenSSL but [tries to load it during runtime](https://github.com/facebookincubator/gloo/blob/2e0fe467aac7ed96a829548ec917edf6be8b1d84/gloo/transport/tcp/tls/openssl.cc#L4-L16)

@malfet @seemethere **I assume that during the release all CI release Linux environments will have OpenSSL 1.1.1(openssl-devel/libssl-dev) installed and it will automatically build PyTorch with TLS_TCP transport support.** 

**The remaining issue: mobile builds and libtorch builds have OpenSSL 1.0.2 which is not compatible with my code**

Depens on https://github.com/pytorch/builder/pull/708


Differential Revision: [D27896285](https://our.internmc.facebook.com/intern/diff/D27896285)

[ghstack-poisoned]

pbelevich committed:
Update on "Add Gloo TCP_TLS transport"

This PR makes the following changes to root CMakeLists.txt:
1. If `USE_GLOO` is `ON` then try to build Gloo with OpenSSL headers(!) and set `USE_GLOO_WITH_OPENSSL` to `ON`
2. If `USE_GLOO_WITH_OPENSSL` is `ON` and the OS is Linux then try to find OpenSSL 1.1.x
3. If `OPENSSL_FOUND` then set `USE_TCP_OPENSSL_LOAD` to `ON` which means that Gloo will use OpenSSL headers to build TLS_TCP transport that is neither linked statically nor dynamically to OpenSSL but [tries to load it during runtime](https://github.com/facebookincubator/gloo/blob/2e0fe467aac7ed96a829548ec917edf6be8b1d84/gloo/transport/tcp/tls/openssl.cc#L4-L16)

@malfet @seemethere **I assume that during the release all CI release Linux environments will have OpenSSL 1.1.1(openssl-devel/libssl-dev) installed and it will automatically build PyTorch with TLS_TCP transport support.** 

**The remaining issue: mobile builds and libtorch builds have OpenSSL 1.0.2 which is not compatible with my code**

Depens on https://github.com/pytorch/builder/pull/708


Differential Revision: [D27896285](https://our.internmc.facebook.com/intern/diff/D27896285)

[ghstack-poisoned]

pbelevich committed:
Update on "Add Gloo TCP_TLS transport"

This PR makes the following changes to root CMakeLists.txt:
1. If `USE_GLOO` is `ON` then try to build Gloo with OpenSSL headers(!) and set `USE_GLOO_WITH_OPENSSL` to `ON`
2. If `USE_GLOO_WITH_OPENSSL` is `ON` and the OS is Linux then try to find OpenSSL 1.1.x
3. If `OPENSSL_FOUND` then set `USE_TCP_OPENSSL_LOAD` to `ON` which means that Gloo will use OpenSSL headers to build TLS_TCP transport that is neither linked statically nor dynamically to OpenSSL but [tries to load it during runtime](https://github.com/facebookincubator/gloo/blob/2e0fe467aac7ed96a829548ec917edf6be8b1d84/gloo/transport/tcp/tls/openssl.cc#L4-L16)

@malfet @seemethere **I assume that during the release all CI release Linux environments will have OpenSSL 1.1.1(openssl-devel/libssl-dev) installed and it will automatically build PyTorch with TLS_TCP transport support.** 

**The remaining issue: mobile builds and libtorch builds have OpenSSL 1.0.2 which is not compatible with my code**

Depens on https://github.com/pytorch/builder/pull/708


Differential Revision: [D27896285](https://our.internmc.facebook.com/intern/diff/D27896285)

[ghstack-poisoned]

pbelevich committed:
Update on "Add Gloo TCP_TLS transport"

This PR makes the following changes to root CMakeLists.txt:
1. If `USE_GLOO` is `ON` then try to build Gloo with OpenSSL headers(!) and set `USE_GLOO_WITH_OPENSSL` to `ON`
2. If `USE_GLOO_WITH_OPENSSL` is `ON` and the OS is Linux then try to find OpenSSL 1.1.x
3. If `OPENSSL_FOUND` then set `USE_TCP_OPENSSL_LOAD` to `ON` which means that Gloo will use OpenSSL headers to build TLS_TCP transport that is neither linked statically nor dynamically to OpenSSL but [tries to load it during runtime](https://github.com/facebookincubator/gloo/blob/2e0fe467aac7ed96a829548ec917edf6be8b1d84/gloo/transport/tcp/tls/openssl.cc#L4-L16)

@malfet @seemethere **I assume that during the release all CI release Linux environments will have OpenSSL 1.1.1(openssl-devel/libssl-dev) installed and it will automatically build PyTorch with TLS_TCP transport support.** 

**The remaining issue: mobile builds and libtorch builds have OpenSSL 1.0.2 which is not compatible with my code**

Depens on https://github.com/pytorch/builder/pull/708


Differential Revision: [D27896285](https://our.internmc.facebook.com/intern/diff/D27896285)

[ghstack-poisoned]

pbelevich committed:
Update on "Add Gloo TCP_TLS transport"

This PR makes the following changes to root CMakeLists.txt:
1. If `USE_GLOO` is `ON` then try to build Gloo with OpenSSL headers(!) and set `USE_GLOO_WITH_OPENSSL` to `ON`
2. If `USE_GLOO_WITH_OPENSSL` is `ON` and the OS is Linux then try to find OpenSSL 1.1.x
3. If `OPENSSL_FOUND` then set `USE_TCP_OPENSSL_LOAD` to `ON` which means that Gloo will use OpenSSL headers to build TLS_TCP transport that is neither linked statically nor dynamically to OpenSSL but [tries to load it during runtime](https://github.com/facebookincubator/gloo/blob/2e0fe467aac7ed96a829548ec917edf6be8b1d84/gloo/transport/tcp/tls/openssl.cc#L4-L16)

@malfet @seemethere **I assume that during the release all CI release Linux environments will have OpenSSL 1.1.1(openssl-devel/libssl-dev) installed and it will automatically build PyTorch with TLS_TCP transport support.** 

**The remaining issue: mobile builds and libtorch builds have OpenSSL 1.0.2 which is not compatible with my code**

Depens on https://github.com/pytorch/builder/pull/708


Differential Revision: [D27896285](https://our.internmc.facebook.com/intern/diff/D27896285)

[ghstack-poisoned]

pbelevich committed:
Update on "Add Gloo TCP_TLS transport"

This PR makes the following changes to root CMakeLists.txt:
1. If `USE_GLOO` is `ON` then try to build Gloo with OpenSSL headers(!) and set `USE_GLOO_WITH_OPENSSL` to `ON`
2. If `USE_GLOO_WITH_OPENSSL` is `ON` and the OS is Linux then try to find OpenSSL 1.1.x
3. If `OPENSSL_FOUND` then set `USE_TCP_OPENSSL_LOAD` to `ON` which means that Gloo will use OpenSSL headers to build TLS_TCP transport that is neither linked statically nor dynamically to OpenSSL but [tries to load it during runtime](https://github.com/facebookincubator/gloo/blob/2e0fe467aac7ed96a829548ec917edf6be8b1d84/gloo/transport/tcp/tls/openssl.cc#L4-L16)

@malfet @seemethere **I assume that during the release all CI release Linux environments will have OpenSSL 1.1.1(openssl-devel/libssl-dev) installed and it will automatically build PyTorch with TLS_TCP transport support.** 

**The remaining issue: mobile builds and libtorch builds have OpenSSL 1.0.2 which is not compatible with my code**

Depens on https://github.com/pytorch/builder/pull/708


Differential Revision: [D27896285](https://our.internmc.facebook.com/intern/diff/D27896285)

[ghstack-poisoned]

pbelevich committed:
Update on "Add Gloo TCP_TLS transport"

This PR makes the following changes to root CMakeLists.txt:
1. If `USE_GLOO` is `ON` then try to build Gloo with OpenSSL headers(!) and set `USE_GLOO_WITH_OPENSSL` to `ON`
2. If `USE_GLOO_WITH_OPENSSL` is `ON` and the OS is Linux then try to find OpenSSL 1.1.x
3. If `OPENSSL_FOUND` then set `USE_TCP_OPENSSL_LOAD` to `ON` which means that Gloo will use OpenSSL headers to build TLS_TCP transport that is neither linked statically nor dynamically to OpenSSL but [tries to load it during runtime](https://github.com/facebookincubator/gloo/blob/2e0fe467aac7ed96a829548ec917edf6be8b1d84/gloo/transport/tcp/tls/openssl.cc#L4-L16)

@malfet @seemethere **I assume that during the release all CI release Linux environments will have OpenSSL 1.1.1(openssl-devel/libssl-dev) installed and it will automatically build PyTorch with TLS_TCP transport support.** 

**The remaining issue: mobile builds and libtorch builds have OpenSSL 1.0.2 which is not compatible with my code**

Depens on https://github.com/pytorch/builder/pull/708


Differential Revision: [D27896285](https://our.internmc.facebook.com/intern/diff/D27896285)

[ghstack-poisoned]

pbelevich committed:
Update on "Add Gloo TCP_TLS transport"

This PR makes the following changes to root CMakeLists.txt:
1. If `USE_GLOO` is `ON` then try to build Gloo with OpenSSL headers(!) and set `USE_GLOO_WITH_OPENSSL` to `ON`
2. If `USE_GLOO_WITH_OPENSSL` is `ON` and the OS is Linux then try to find OpenSSL 1.1.x
3. If `OPENSSL_FOUND` then set `USE_TCP_OPENSSL_LOAD` to `ON` which means that Gloo will use OpenSSL headers to build TLS_TCP transport that is neither linked statically nor dynamically to OpenSSL but [tries to load it during runtime](https://github.com/facebookincubator/gloo/blob/2e0fe467aac7ed96a829548ec917edf6be8b1d84/gloo/transport/tcp/tls/openssl.cc#L4-L16)

@malfet @seemethere **I assume that during the release all CI release Linux environments will have OpenSSL 1.1.1(openssl-devel/libssl-dev) installed and it will automatically build PyTorch with TLS_TCP transport support.** 

**The remaining issue: mobile builds and libtorch builds have OpenSSL 1.0.2 which is not compatible with my code**

Depens on https://github.com/pytorch/builder/pull/708


Differential Revision: [D27896285](https://our.internmc.facebook.com/intern/diff/D27896285)

[ghstack-poisoned]

pbelevich committed:
Update on "Add Gloo TCP_TLS transport"

This PR makes the following changes to root CMakeLists.txt:
1. If `USE_GLOO` is `ON` then try to build Gloo with OpenSSL headers(!) and set `USE_GLOO_WITH_OPENSSL` to `ON`
2. If `USE_GLOO_WITH_OPENSSL` is `ON` and the OS is Linux then try to find OpenSSL 1.1.x
3. If `OPENSSL_FOUND` then set `USE_TCP_OPENSSL_LOAD` to `ON` which means that Gloo will use OpenSSL headers to build TLS_TCP transport that is neither linked statically nor dynamically to OpenSSL but [tries to load it during runtime](https://github.com/facebookincubator/gloo/blob/2e0fe467aac7ed96a829548ec917edf6be8b1d84/gloo/transport/tcp/tls/openssl.cc#L4-L16)

@malfet @seemethere **I assume that during the release all CI release Linux environments will have OpenSSL 1.1.1(openssl-devel/libssl-dev) installed and it will automatically build PyTorch with TLS_TCP transport support.** 

**The remaining issue: mobile builds and libtorch builds have OpenSSL 1.0.2 which is not compatible with my code**

Depens on https://github.com/pytorch/builder/pull/708


Differential Revision: [D27896285](https://our.internmc.facebook.com/intern/diff/D27896285)

[ghstack-poisoned]

pbelevich committed:
Update on "Add Gloo TCP_TLS transport"

This PR makes the following changes to root CMakeLists.txt:
1. If `USE_GLOO` is `ON` then try to build Gloo with OpenSSL headers(!) and set `USE_GLOO_WITH_OPENSSL` to `ON`
2. If `USE_GLOO_WITH_OPENSSL` is `ON` and the OS is Linux then try to find OpenSSL 1.1.x
3. If `OPENSSL_FOUND` then set `USE_TCP_OPENSSL_LOAD` to `ON` which means that Gloo will use OpenSSL headers to build TLS_TCP transport that is neither linked statically nor dynamically to OpenSSL but [tries to load it during runtime](https://github.com/facebookincubator/gloo/blob/2e0fe467aac7ed96a829548ec917edf6be8b1d84/gloo/transport/tcp/tls/openssl.cc#L4-L16)

@malfet @seemethere **I assume that during the release all CI release Linux environments will have OpenSSL 1.1.1(openssl-devel/libssl-dev) installed and it will automatically build PyTorch with TLS_TCP transport support.** 

**The remaining issue: mobile builds and libtorch builds have OpenSSL 1.0.2 which is not compatible with my code**

Depens on https://github.com/pytorch/builder/pull/708


Differential Revision: [D27896285](https://our.internmc.facebook.com/intern/diff/D27896285)

[ghstack-poisoned]

pbelevich committed:
Update on "Add Gloo TCP_TLS transport"

This PR makes the following changes to root CMakeLists.txt:
1. If `USE_GLOO` is `ON` then try to build Gloo with OpenSSL headers(!) and set `USE_GLOO_WITH_OPENSSL` to `ON`
2. If `USE_GLOO_WITH_OPENSSL` is `ON` and the OS is Linux then try to find OpenSSL 1.1.x
3. If `OPENSSL_FOUND` then set `USE_TCP_OPENSSL_LOAD` to `ON` which means that Gloo will use OpenSSL headers to build TLS_TCP transport that is neither linked statically nor dynamically to OpenSSL but [tries to load it during runtime](https://github.com/facebookincubator/gloo/blob/2e0fe467aac7ed96a829548ec917edf6be8b1d84/gloo/transport/tcp/tls/openssl.cc#L4-L16)

@malfet @seemethere **I assume that during the release all CI release Linux environments will have OpenSSL 1.1.1(openssl-devel/libssl-dev) installed and it will automatically build PyTorch with TLS_TCP transport support.** 

**The remaining issue: mobile builds and libtorch builds have OpenSSL 1.0.2 which is not compatible with my code**

Depens on https://github.com/pytorch/builder/pull/708


Differential Revision: [D27896285](https://our.internmc.facebook.com/intern/diff/D27896285)

[ghstack-poisoned]

pbelevich committed:
Update on "Add Gloo TCP_TLS transport"

This PR makes the following changes to root CMakeLists.txt:
1. If `USE_GLOO` is `ON` then try to build Gloo with OpenSSL headers(!) and set `USE_GLOO_WITH_OPENSSL` to `ON`
2. If `USE_GLOO_WITH_OPENSSL` is `ON` and the OS is Linux then try to find OpenSSL 1.1.x
3. If `OPENSSL_FOUND` then set `USE_TCP_OPENSSL_LOAD` to `ON` which means that Gloo will use OpenSSL headers to build TLS_TCP transport that is neither linked statically nor dynamically to OpenSSL but [tries to load it during runtime](https://github.com/facebookincubator/gloo/blob/2e0fe467aac7ed96a829548ec917edf6be8b1d84/gloo/transport/tcp/tls/openssl.cc#L4-L16)

@malfet @seemethere **I assume that during the release all CI release Linux environments will have OpenSSL 1.1.1(openssl-devel/libssl-dev) installed and it will automatically build PyTorch with TLS_TCP transport support.** 

**The remaining issue: mobile builds and libtorch builds have OpenSSL 1.0.2 which is not compatible with my code**

Depens on https://github.com/pytorch/builder/pull/708


Differential Revision: [D27896285](https://our.internmc.facebook.com/intern/diff/D27896285)

[ghstack-poisoned]

pbelevich committed:
Update on "Add Gloo TCP_TLS transport"

This PR makes the following changes to root CMakeLists.txt:
1. If `USE_GLOO` is `ON` then try to build Gloo with OpenSSL headers(!) and set `USE_GLOO_WITH_OPENSSL` to `ON`
2. If `USE_GLOO_WITH_OPENSSL` is `ON` and the OS is Linux then try to find OpenSSL 1.1.x
3. If `OPENSSL_FOUND` then set `USE_TCP_OPENSSL_LOAD` to `ON` which means that Gloo will use OpenSSL headers to build TLS_TCP transport that is neither linked statically nor dynamically to OpenSSL but [tries to load it during runtime](https://github.com/facebookincubator/gloo/blob/2e0fe467aac7ed96a829548ec917edf6be8b1d84/gloo/transport/tcp/tls/openssl.cc#L4-L16)

@malfet @seemethere **I assume that during the release all CI release Linux environments will have OpenSSL 1.1.1(openssl-devel/libssl-dev) installed and it will automatically build PyTorch with TLS_TCP transport support.** 

**The remaining issue: mobile builds and libtorch builds have OpenSSL 1.0.2 which is not compatible with my code**

Depens on https://github.com/pytorch/builder/pull/708


Differential Revision: [D27896285](https://our.internmc.facebook.com/intern/diff/D27896285)

[ghstack-poisoned]

pbelevich committed:
Update on "Add Gloo TCP_TLS transport"

This PR makes the following changes to root CMakeLists.txt:
1. If `USE_GLOO` is `ON` then try to build Gloo with OpenSSL headers(!) and set `USE_GLOO_WITH_OPENSSL` to `ON`
2. If `USE_GLOO_WITH_OPENSSL` is `ON` and the OS is Linux then try to find OpenSSL 1.1.x
3. If `OPENSSL_FOUND` then set `USE_TCP_OPENSSL_LOAD` to `ON` which means that Gloo will use OpenSSL headers to build TLS_TCP transport that is neither linked statically nor dynamically to OpenSSL but [tries to load it during runtime](https://github.com/facebookincubator/gloo/blob/2e0fe467aac7ed96a829548ec917edf6be8b1d84/gloo/transport/tcp/tls/openssl.cc#L4-L16)

@malfet @seemethere **I assume that during the release all CI release Linux environments will have OpenSSL 1.1.1(openssl-devel/libssl-dev) installed and it will automatically build PyTorch with TLS_TCP transport support.** 

**The remaining issue: mobile builds and libtorch builds have OpenSSL 1.0.2 which is not compatible with my code**

Depens on https://github.com/pytorch/builder/pull/708


Differential Revision: [D27896285](https://our.internmc.facebook.com/intern/diff/D27896285)

[ghstack-poisoned]

pbelevich commented:
1. `find_package(OpenSSL REQUIRED)` removed from pytorch CMakeLists.txt
2. `USE_GLOO_WITH_OPENSSL` is OFF by default
3. Generating test cert/key in bash script was replaced with python code (create_test_cert.py), but I still have create_test_cert.sh to run it
4. I limited running TLS transport tests to `pytorch-linux-xenial-py3.6-gcc7-test` and `pytorch-linux-xenial-py3.6-gcc5.4-test`
5. I limited running TLS transport tests to only 44 tests in `ProcessGroupGlooTest` in `test_c10d_gloo.py`, it takes 20 seconds to run
6. I didn't add a test that checks that setting `GLOO_DEVICE_TRANSPORT` to `FOOBAR` raises an exception, because it requires some extra work, but I guarantee that `GLOO_DEVICE_TRANSPORT=FOOBAR` will crash with seg-fault(as expected)

pbelevich commented:
@pbelevich has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27896285).

pbelevich committed:
Update on "Add Gloo TCP_TLS transport"

This PR makes the following changes to root CMakeLists.txt:
1. If `USE_GLOO` is `ON` then try to build Gloo with OpenSSL headers(!) and set `USE_GLOO_WITH_OPENSSL` to `ON`
2. If `USE_GLOO_WITH_OPENSSL` is `ON` and the OS is Linux then try to find OpenSSL 1.1.x
3. If `OPENSSL_FOUND` then set `USE_TCP_OPENSSL_LOAD` to `ON` which means that Gloo will use OpenSSL headers to build TLS_TCP transport that is neither linked statically nor dynamically to OpenSSL but [tries to load it during runtime](https://github.com/facebookincubator/gloo/blob/2e0fe467aac7ed96a829548ec917edf6be8b1d84/gloo/transport/tcp/tls/openssl.cc#L4-L16)

@malfet @seemethere **I assume that during the release all CI release Linux environments will have OpenSSL 1.1.1(openssl-devel/libssl-dev) installed and it will automatically build PyTorch with TLS_TCP transport support.** 

**The remaining issue: mobile builds and libtorch builds have OpenSSL 1.0.2 which is not compatible with my code**

Depens on https://github.com/pytorch/builder/pull/708


Differential Revision: [D27896285](https://our.internmc.facebook.com/intern/diff/D27896285)

[ghstack-poisoned]

pbelevich commented:
@pbelevich has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27896285).

pbelevich committed:
Update on "Add Gloo TCP_TLS transport"

This PR makes the following changes to root CMakeLists.txt:
1. If `USE_GLOO` is `ON` then try to build Gloo with OpenSSL headers(!) and set `USE_GLOO_WITH_OPENSSL` to `ON`
2. If `USE_GLOO_WITH_OPENSSL` is `ON` and the OS is Linux then try to find OpenSSL 1.1.x
3. If `OPENSSL_FOUND` then set `USE_TCP_OPENSSL_LOAD` to `ON` which means that Gloo will use OpenSSL headers to build TLS_TCP transport that is neither linked statically nor dynamically to OpenSSL but [tries to load it during runtime](https://github.com/facebookincubator/gloo/blob/2e0fe467aac7ed96a829548ec917edf6be8b1d84/gloo/transport/tcp/tls/openssl.cc#L4-L16)

@malfet @seemethere **I assume that during the release all CI release Linux environments will have OpenSSL 1.1.1(openssl-devel/libssl-dev) installed and it will automatically build PyTorch with TLS_TCP transport support.** 

**The remaining issue: mobile builds and libtorch builds have OpenSSL 1.0.2 which is not compatible with my code**

Depens on https://github.com/pytorch/builder/pull/708


Differential Revision: [D27896285](https://our.internmc.facebook.com/intern/diff/D27896285)

[ghstack-poisoned]

pbelevich commented:
@pbelevich has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27896285).

pbelevich committed:
Update on "Add Gloo TCP_TLS transport"

This PR makes the following changes to root CMakeLists.txt:
1. If `USE_GLOO` is `ON` then try to build Gloo with OpenSSL headers(!) and set `USE_GLOO_WITH_OPENSSL` to `ON`
2. If `USE_GLOO_WITH_OPENSSL` is `ON` and the OS is Linux then try to find OpenSSL 1.1.x
3. If `OPENSSL_FOUND` then set `USE_TCP_OPENSSL_LOAD` to `ON` which means that Gloo will use OpenSSL headers to build TLS_TCP transport that is neither linked statically nor dynamically to OpenSSL but [tries to load it during runtime](https://github.com/facebookincubator/gloo/blob/2e0fe467aac7ed96a829548ec917edf6be8b1d84/gloo/transport/tcp/tls/openssl.cc#L4-L16)

@malfet @seemethere **I assume that during the release all CI release Linux environments will have OpenSSL 1.1.1(openssl-devel/libssl-dev) installed and it will automatically build PyTorch with TLS_TCP transport support.** 

**The remaining issue: mobile builds and libtorch builds have OpenSSL 1.0.2 which is not compatible with my code**

Depens on https://github.com/pytorch/builder/pull/708


Differential Revision: [D27896285](https://our.internmc.facebook.com/intern/diff/D27896285)

[ghstack-poisoned]

pbelevich committed:
Update on "Add Gloo TCP_TLS transport"

This PR makes the following changes to root CMakeLists.txt:
1. If `USE_GLOO` is `ON` then try to build Gloo with OpenSSL headers(!) and set `USE_GLOO_WITH_OPENSSL` to `ON`
2. If `USE_GLOO_WITH_OPENSSL` is `ON` and the OS is Linux then try to find OpenSSL 1.1.x
3. If `OPENSSL_FOUND` then set `USE_TCP_OPENSSL_LOAD` to `ON` which means that Gloo will use OpenSSL headers to build TLS_TCP transport that is neither linked statically nor dynamically to OpenSSL but [tries to load it during runtime](https://github.com/facebookincubator/gloo/blob/2e0fe467aac7ed96a829548ec917edf6be8b1d84/gloo/transport/tcp/tls/openssl.cc#L4-L16)

@malfet @seemethere **I assume that during the release all CI release Linux environments will have OpenSSL 1.1.1(openssl-devel/libssl-dev) installed and it will automatically build PyTorch with TLS_TCP transport support.** 

**The remaining issue: mobile builds and libtorch builds have OpenSSL 1.0.2 which is not compatible with my code**

Depens on https://github.com/pytorch/builder/pull/708


Differential Revision: [D27896285](https://our.internmc.facebook.com/intern/diff/D27896285)

[ghstack-poisoned]

pbelevich commented:
@pbelevich has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27896285).
